clusterspec/clusterspec.go

```diff
@@ -0,0 +1,133 @@
+package clusterspec
```

Comment:
Why another top level directory now? I think this should either be in `service` or `pkg`, which are both the same for me personally so I still do not understand why we have both in the first place.

Reply:
`pkg` is only for implementation that can be exported and meant to be used also from repositories outside this repo. Examples can be found from Kubernetes, Docker, Rkt, CockroachDB and numerous other Go projects.

I feel `service`  is inappropriate for two reasons:
- In general in Go one should avoid deeply nested packages. They are ok from time to time, but they don't work like in other languages (such as Java for example).
- This is not a service implementation or even related. This is to implementation to supplement information from one domain model with information from runtime configuration and then construct a new instance for another domain model.

Reply:
I'm not expecting code under `pkg` to be necessarily used from outside, but it's a way to communicate that code in there is not tightly coupled with implementation in this project and e.g. `certconfig` resource in here is a good example of this case. This view point also leads to cleaner architecture as can be seen here in form of a requirement for this factory abstraction.

I could be biased in here having long background in web application backends, but for me Service as a term communicates a logical component that provides interface for business logic. The actual business logic then is often implemented elsewhere (e.g. in domain model in DDD), but service implementation composes those implementations as a higher level exportable interface.

How I have seen Go "coding culture" it is that one does not embrace services so much but concentrate on domain model, package code accordingly and then accompany it with code and then compose these together. There's also a Golang blog post related to this: https://blog.golang.org/organizing-go-code

Reply:
IMO this is what we are doing. It is only that we bundle the code in the service package, which is now broken anyway. I let this go now.